### PR TITLE
Fix `items_after_test_module` for non root modules, add applicable suggestion

### LINKED ISSUE #11050: items_after_test_module interacts badly with rstest and submodules
### Summary

I like to organize my tests into multiple modules just to provide namespacing.  For example, I might use one module for each function under test, and each of those modules will contain one function for each test case, like this:
```rust
mod func1 {
    use super::*;
    #[test]
    fn test_case1() {...}
    #[test]
    fn test_case2() {...}
}
mod func2 {
    use super::*;
    #[test]
    fn test_case1() {...}
    #[test]
    fn test_case2() {...}
}
```
The problem is that if anything above the final module includes a `#[cfg(test)]`, then the items_after_test_module lint will fire.  That's even true in integration tests, which aren't built with and without `#[cfg(test)]`.  The problem is exacerbated by macro crates like `rstest`, which sometimes emit a `#[cfg(test)]` in the generated code.  And it's frustrating to debug, because the user can't even see the offending `#[cfg(test)]`.
I don't know what a perfect solution would be.  But skipping this lint in integration tests would be a good start.  Or, skipping this lint within an outer module which is itself `#[cfg(test)]`.

### Lint Name

items_after_test_module

### Reproducer

I tried this code:

```rust
use rstest::rstest;

#[rstest]
#[case(42)]
fn t(#[case] _x: u32) {}

mod func1 {
    struct SomethingUsedByTheTest{}

    #[test]
    fn t() {}
}
```

I saw this happen:

```
warning: items were found after the testing module
  --> tests/t.rs:3:1
   |
3  |   #[rstest]
   |   ^--------
   |   |
   |  _in this procedural macro expansion
   | |
4  | | #[case(42)]
5  | | fn t(#[case] _x: u32) {}
6  | |
...  |
10 | |     #[test]
11 | |     fn t() {}
   | |_____________^
   |
   = help: move the items to before the testing module was defined
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#items_after_test_module
   = note: `#[warn(clippy::items_after_test_module)]` on by default
   = note: this warning originates in the attribute macro `rstest` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I expected to see this happen:

No errors (except for `SomethingUsedByTheTest is never constructed`, of course).

### Version

```text
rustc 1.72.0-nightly (5bd28f5ea 2023-06-28)
binary: rustc
commit-hash: 5bd28f5eac1ba3569bfa8d49ec3f5acbdfdff7a0
commit-date: 2023-06-28
host: x86_64-unknown-freebsd
release: 1.72.0-nightly
LLVM version: 16.0.5
```

### Additional Labels

_No response_

### LINKED ISSUE #11153: `items-after-test-module`: false positive with `test-case` and regular tests
### Summary

https://github.com/rust-lang/rust-clippy/pull/10992 fixed some of the interaction between the `items-after-test-module` lint and the `test-case` crate by ignoring proc macros. That works, but it still gets confused if you have non-`test-case` tests after `test-case` tests in the same file.

### Lint Name

items-after-test-module

### Reproducer

I tried this code:

```rust
use test_case::test_case;

#[test_case(1)]
#[test_case(2)]
fn test_the_thing(n: usize) {
    assert!(n < 5);
}

#[test]
fn other_test() {
    assert!(5 < 6);
}
```

I saw this happen:

```
warning: items were found after the testing module
  --> tests/stuff.rs:3:1
   |
3  |   #[test_case(1)]
   |   ^--------------
   |   |
   |  _in this procedural macro expansion
   | |
4  | | #[test_case(2)]
5  | | fn test_the_thing(n: usize) {
6  | |     assert!(n < 5);
...  |
11 | |     assert!(5 < 6);
12 | | }
   | |_^
   |
   = help: move the items to before the testing module was defined
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#items_after_test_module
   = note: `-W clippy::items-after-test-module` implied by `-W clippy::all`
   = note: this warning originates in the attribute macro `test_case` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Version

```text
rustc 1.71.0 (8ede3aae2 2023-07-12)
binary: rustc
commit-hash: 8ede3aae28fe6e4d52b38157d7bfe0d3bceef225
commit-date: 2023-07-12
host: aarch64-apple-darwin
release: 1.71.0
LLVM version: 16.0.5
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by Alexendoo**
Fixes #11050
Fixes #11153

changelog: [`items_after_test_module`]: Now suggests a machine-applicable suggestion.
changelog: [`items:after_test_module`]: Also lints for non root modules

### CODE REVIEW

**Comment by blyxyas** [comment]
I was testing the `master` version of this lint with the following setup:

```
// main.rs ==================================
#![deny(clippy::items_after_test_module)]

mod other_mod;
use crate::other_mod::add;

fn main() {
    println!("{}", add(3, 54));
}

// other_mod.rs ==============================
pub fn add(a: usize, b: usize) -> usize {
    a + b
}

#[cfg(test)]
mod tests {
    #[test]
    fn my_test() {}
}

const MY_ITEM: usize = 0;
```

But it doesn't emit. Turns out that the lint doesn't emit on any module that isn't on the `main.rs` / `lib.rs`. This is a bug.

**Comment by Alexendoo** [comment]
Reworked the lint to cover that, also added a hidden suggestion that can be applied by `--fix` and tweaked the spans while I was at it

File: tests/ui/items_after_test_module/in_submodule.stderr
Comment (blyxyas) [inline tests/ui/items_after_test_module/in_submodule.stderr:1]: Kinda weird how what should be `submodule.rs`'s standard error is actually stored in `in_submodule.stderr`

**Review by blyxyas — APPROVED** [review]
Just some mini-nits and a question and this should be ready (｡･ω･｡)

File: tests/ui/items_after_test_module/in_submodule.stderr
Comment (Alexendoo) [reply tests/ui/items_after_test_module/in_submodule.stderr:1]: It's because it's the stderr output from running the `in_submodule.rs` test

**Pull request merged**